boa/src/builtins/number/mod.rs

```diff
@@ -159,6 +159,145 @@ pub fn to_precision(this: &mut Value, args: &[Value], _ctx: &mut Interpreter) ->
     unimplemented!("TODO: Implement toPrecision");
 }
 
+const BUF_SIZE: usize = 2200;
+
+// https://golang.org/src/math/nextafter.go
+#[inline]
+fn next_after(x: f64, y: f64) -> f64 {
+    if x.is_nan() || y.is_nan() {
+        f64::NAN
+    } else if (x - y).abs() < f64::EPSILON {
+        x
+    } else if x == 0.0 {
+        f64::from_bits(1).copysign(y)
+    } else if y > x || x > 0.0 {
+        f64::from_bits(x.to_bits() + 1)
+    } else {
+        f64::from_bits(x.to_bits() - 1)
+    }
+}
+
+// https://stackoverflow.com/questions/55690397/where-is-the-frexp-function-for-f32-in-rust
+#[inline]
+fn frexp(s: f64) -> (f64, isize) {
+    if 0.0 == s {
+        (s, 0)
+    } else {
+        let lg = s.abs().log2();
+        let x = (lg - lg.floor() - 1.0).exp2();
+        let exp = lg.floor() + 1.0;
+        (s.signum() * x, exp as isize)
+    }
+}
+
+// https://chromium.googlesource.com/v8/v8/+/refs/heads/master/src/numbers/conversions.cc#1230
+pub fn num_to_string(mut value: f64, radix: u32) -> String {
+    assert!(radix >= 2);
+    assert!(radix <= 36);
+    assert!(value.is_finite());
+    // assert_ne!(0.0, value);
+
+    // Character array used for conversion.
+    // Temporary buffer for the result. We start with the decimal point in the
+    // middle and write to the left for the integer part and to the right for the
+    // fractional part. 1024 characters for the exponent and 52 for the mantissa
+    // either way, with additional space for sign, decimal point and string
+    // termination should be sufficient.
+    let mut buffer: [u8; BUF_SIZE] = [0; BUF_SIZE];
+    let mut integer_cursor = BUF_SIZE / 2;
+    let mut fraction_cursor = integer_cursor;
+    // let negative: bool = value < 0f64;
+    let negative = value.is_sign_negative();
+    if negative {
+        value = -value
+    }
+    // Split the value into an integer part and a fractional part.
+    let mut integer = value.trunc();
+    let mut fraction = value.fract();
+
+    // We only compute fractional digits up to the input double's precision.
+    let mut delta = 0.5 * (next_after(value, f64::MAX) - value);
+    delta = next_after(0.0, f64::MAX).max(delta);
+    assert!(delta > 0.0);
+    if fraction >= delta {
+        // Insert decimal point.
+        let fresh0 = fraction_cursor;
+        fraction_cursor += 1;
+        buffer[fresh0] = b'.';
```

Comment:
In Rust, this indexing is **very** slow, since it will do bounds checking every time it accesses the buffer. I think we can somehow iterate through it using iterators instead of direct access, but then, the carry-overs can be problematic. Maybe @HalidOdat has some ideas on how to improve this performance.

Reply:
Something that can be done here to at least increase the iteration for the integer part is this:
```rust
let (int_buf, frac_buf) = buffer.split_at_mut(BUF_SIZE / 2);
```

Then, in the integer iteration:
```rust
    let mut int_iter = int_buf.into_iter().rev();
    for c in int_iter {
        let remainder = integer % (radix as f64);
        *c = std::char::from_digit(remainder as u32, radix).unwrap() as u8;
        integer = (integer - remainder) / radix as f64;
        if integer <= 0_f64 {
            break;
        }
    }
    // Add sign and terminate string.
    if negative {
        *int_buf.next().expect("buffer ended") = b'-';
    }
```

This might need some `mut`-s somewhere around, and we need to save the index somehow, maybe by calling `enumerate()` in the iterator, but it should work. Maybe we can do something similar with the fractional part.

Reply:
We may also be able to replace some of the while loop using `take_while`.

Reply:
Fraction part is hard to refactor since there is a second pass to the buffer and implementation sometimes increments and sometimes decrements the cursor. Maybe we can do something with `Rev` but it will be complicated.

Reply:
Something like bidir_iter might work better but problem is they use slice.get internally so we still pay for the bound checks. We might use unsafe get to not pay for the bound check since we check the bounds anyway.

Reply:
I think it's good enough for now :) we can maybe improve it further in the future.